building-and-packaging/material/cmake.md

```diff
@@ -0,0 +1,414 @@
+- `mkdir clean` and `make clean` confuses Make
+
+```
++ .PHONY : all clean
+```
+
+### Advanced Make
+
+- There is more:
+    - Variables
+    - Rules
+    - Wildcards
+- ... but becomes quickly very hard to read
+- Not covered because cmake does this for us.
+- But nicely documented in [py-RSE chapter 9](https://merely-useful.tech/py-rse/automate.html).
+
+
+## CMake
+
+*35 mins* ... might be tough
+
+### Introduction
+
+
+What is cmake?
+
+> A build system generator, for e.g. makefiles
+
+Why not directly use makefiles?
+
+> CMake can do many more things, is more flexible, avoids complex patterns, and is, thus, more readable
+
+Are there alternatives?
+
+> Yes, `scons`, `autotools`, ...
+
+Will you explain and discuss the differences in detail?
+
+> No, but I claim that CMake is the most standard tool and sticking to standards has a value in itself. [Watch a video why](https://youtu.be/sBP17HQAQjk)
+
+Can you give some reasons why CMake is great?
+
+> Of course:
+
+- CMake can generate many build systems: make, ninja, VSCode project, Eclipse project, ...
+- Many GUIs and TUIs: ccmake, cmake-gui, integration in probably nearly all IDEs, ...
+- CMake is *"cross-platform"*: you can ideally use same cmake file in all OS's (easy to distinguish platform-specific things)
+- Build directory becomes independent of source directory (build multiple versions with different dependencies etc.)
+- CMake respects choices in user environment (e.g. user defines cpp compiler through CXX)
+
+### CMake "Hello World"
+
+Example adapted from [Yanson Tech](https://www.youtube.com/watch?v=wl2Srog-j7I)
+Go back to original `main.cpp`
+
+`CMakeLists.txt`:
+
+```cmake
+cmake_minimum_required(VERSION "3.12")
+
+project("HelloWorld")
+
+add_executable("${PROJECT_NAME}" main.cpp)
+```
+
+- CMake variable values: `${}`, use quotations to avoid problems with spaces
+- `mkdir build && cd build && cmake ..` ... explain files (TODO), look at makefile
+    - `CMakeCache.txt`
+- `make` and `./HelloWorld`
+
+### Multiple files
+
+- Add again `sse.cpp` and `sse.hpp` in subfolder `sse` and adapt `main.cpp`
+
+`CMakeLists.txt`:
+
+```diff
++ add_executable("${PROJECT_NAME}" main.cpp sse/sse.cpp)
+- add_executable("${PROJECT_NAME}" main.cpp)
+```
+
+- Generate and build again, this time both with verbose output TODO
+
+- What if there are multiple source files in `sse` folder?
+
+`CMakeLists.txt`:
+
+```diff
++ file(GLOB_RECURSE SRC_FILES CONFIGURE_DEPENDS main.cpp sse/*.cpp)
++ add_executable("${PROJECT_NAME}" "${SRC_FILES}")
+- add_executable("${PROJECT_NAME}" main.cpp sse.cpp)
+```
+
+- `GLOB_RECURSE`: filenames with wildcard characters, recurse -> also subdirectories
+- Actually discouraged since wildcard populated while generating. You need to run `cmake` again when you add a file. See [cmake docs](https://cmake.org/cmake/help/latest/command/file.html#filesystem).
+- `CONFIGURE_DEPENDS` partial remedy, but costly and no guarantee that all build systems support this
```

Comment:
So is there a proper way around it? Should one add source files by hand?

Reply:
A proper way around is to generate the list of source files outside of the main cmake file. At least, this is how it is done in preCICE.